halo2_proofs/src/plonk/circuit.rs

```diff
@@ -1352,6 +1353,34 @@ impl<F: Field> Mul<F> for Expression<F> {
     }
 }
 
+impl<F: Field> Sum<Self> for Expression<F> {
+    fn sum<I: Iterator<Item = Self>>(mut iter: I) -> Self {
+        if let Some(mut result) = iter.next() {
+            for term in iter {
+                result = result + term;
+            }
+
+            result
+        } else {
+            Self::Constant(F::ZERO)
+        }
```

Comment:
Why not using `iter.fold()` here which doesn't require any `if else` clauses??
I'd re-do this.

Reply:
If I used `fold`, `vec![1, 2].into_iter().fold(F::ZERO, |acc, x| acc + x)` would become `Expression::Sum(Expression::Sum(Expression::Sum(0, 1), 2), 3)`. Unnecessary 0 term happens. I tried to avoid this unnecessary 0 term for optimization.

Reply:
You have to create the constant anyways. As the expressions need to be hardcoded as constants within the `ConstraintSystem` I'd say.

The loss we get from the `0` i thin is probably always less than the fact of having `if_else` even with branch prediction at LLVM-level & shenanigans.

@han0110 do you know if the conditional generation of  constant Expressions incurrs into a different circuit VK??

Reply:
I think using if-else here is totally okay, but using `fold` or `reduce` definitely looks better. So if we want to avoid the `0` in `Sum`, we might do:

```rust
iter.reduce(|acc, expr| acc + expr).unwrap_or(Expression::Constant(F::ZERO))
```

Reply:
@chokermaxx I think Han's proposal is what should definitely go in!
Thxz Han!

Reply:
@han0110 this is nice. thanks!